include/RAJA/policy/atomic_builtin.hpp

```diff
@@ -37,389 +44,828 @@ namespace RAJA
 {
-  builtin_atomic_store((unsigned long long *)acc, RAJA::util::reinterp_A_as_B<T, unsigned long long>(value));
+  return RAJA::util::reinterp_A_as_B<builtin_useReinterpret_t<T>, T>(
+    builtin_atomicCAS(
+      reinterpret_cast<builtin_useReinterpret_t<T>*>(acc),
+      RAJA::util::reinterp_A_as_B<T, builtin_useReinterpret_t<T>>(compare),
+      RAJA::util::reinterp_A_as_B<T, builtin_useReinterpret_t<T>>(value)));
 }
 
-template <typename T>
-RAJA_DEVICE_HIP RAJA_INLINE
-    typename std::enable_if<sizeof(T) == sizeof(unsigned), T>::type
-    builtin_atomic_CAS(T *acc, T compare, T value)
+
+/*!
+ * Implementation of compare and swap loop
+ */
+
+
+/*!
+ * Equality comparison for compare and swap loop using types supported by
+ * intrinsics.
+ */
+template <typename T,
+          std::enable_if_t<builtin_useIntrinsic<T>::value, bool> = true>
+RAJA_DEVICE_HIP RAJA_INLINE bool builtin_atomicCAS_equal(const T &a, const T &b)
 {
-  return RAJA::util::reinterp_A_as_B<unsigned, T>(
-      builtin_atomic_CAS((unsigned *)acc,
-                         RAJA::util::reinterp_A_as_B<T, unsigned>(compare),
-                         RAJA::util::reinterp_A_as_B<T, unsigned>(value)));
+  return a == b;
 }
 
-template <typename T>
-RAJA_DEVICE_HIP RAJA_INLINE
-    typename std::enable_if<sizeof(T) == sizeof(unsigned long long), T>::type
-    builtin_atomic_CAS(T *acc, T compare, T value)
+
+/*!
+ * Equality comparison for compare and swap loop using reinterpret cast.
+ * Converts to the underlying integral type to avoid cases where the values
+ * will never compare equal (most notably, NaNs).
+ */
+template <typename T,
+          std::enable_if_t<builtin_useReinterpret<T>::value, bool> = true>
+RAJA_DEVICE_HIP RAJA_INLINE bool builtin_atomicCAS_equal(const T &a, const T &b)
 {
-  return RAJA::util::reinterp_A_as_B<unsigned long long, T>(builtin_atomic_CAS(
-      (unsigned long long *)acc,
-      RAJA::util::reinterp_A_as_B<T, unsigned long long>(compare),
-      RAJA::util::reinterp_A_as_B<T, unsigned long long>(value)));
+  return builtin_atomicCAS_equal(
+    RAJA::util::reinterp_A_as_B<T, builtin_useReinterpret_t<T>>(a),
+    RAJA::util::reinterp_A_as_B<T, builtin_useReinterpret_t<T>>(b));
 }
 
+
 /*!
- * Generic impementation of any atomic 32-bit or 64-bit operator that can be
- * implemented using a compare and swap primitive.
- * Implementation uses the builtin unsigned 32-bit and 64-bit CAS operators.
+ * Generic impementation of any atomic 8, 16, 32, or 64 bit operator
+ * that can be implemented using a builtin compare and swap primitive.
  * Returns the OLD value that was replaced by the result of this operation.
  */
-template <typename T, typename OPER>
-RAJA_DEVICE_HIP RAJA_INLINE T builtin_atomic_CAS_oper(T *acc,
-                                                      OPER &&oper)
+template <typename T, typename Oper>
+RAJA_DEVICE_HIP RAJA_INLINE T builtin_atomicCAS(T *acc, Oper &&oper)
```

Comment:
I'm not sure its a good thing that this and the normal builtin_atomicCAS are overloads.

Reply:
Done.